### PR TITLE
Add 'module-type' option determining how README is linted

### PR DESCRIPTION
**Pull request by Avaq**
# Changes

- Update doctest to the latest version
- Expose doctest' new ESM feature through a `module-type` setting that users can add in their `.config`. This option also ensures that code examples are treated as "modules" by the linter.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -96,6 +96,7 @@ Many variables have default values and are therefore optional.
 | `opening-delimiter`           | `` ```javascript ``   | The opening delimiter of doctest blocks in the source files.                  |
 | `closing-delimiter`           | `` ``` ``             | The closing delimiter of doctest blocks in the source files.                  |
 | `version-tag-prefix`          | `v`                   | The prefix of annotated version tags (`version-tag-prefix =` for no prefix).  |
+| `module-type`                 | `commonjs`            | The type of module doctest should use. One of `commonjs`, `amd`, or `esm`.    |
```
Comment (davidchambers) [inline README.md]: Let's include `module-type` after `closing-delimiter` so the doctest-related variables are together.

Comment (Avaq) [reply README.md]: :zap:

File: bin/lint-readme [hunk 2]
```diff
@@ -68,4 +68,11 @@ node_modules/.bin/remark \
 cp README.md README.md.temp
 rewrite <README.md.temp >README.md
 
-node_modules/.bin/eslint -- README.md
+module="$(get module-type)"
+if [[ $module == 'esm' ]] ; then
+  source='module'
+else
+  source='script'
+fi
```
Comment (davidchambers) [inline bin/lint-readme:76]: All three pairs of single quotes can be dropped. :)

Comment (Avaq) [reply bin/lint-readme:76]: :zap:

File: bin/lint-readme [hunk 3]
```diff
@@ -68,4 +68,11 @@ node_modules/.bin/remark \
 cp README.md README.md.temp
 rewrite <README.md.temp >README.md
 
-node_modules/.bin/eslint -- README.md
+module="$(get module-type)"
```
Comment (davidchambers) [inline bin/lint-readme]: Let's move this alongside the other uses of `get` to make it easy to see which variables this script uses.

Comment (Avaq) [reply bin/lint-readme]: :zap:

File: bin/lint-readme [hunk 4]
```diff
@@ -68,4 +69,10 @@ node_modules/.bin/remark \
...
 rewrite <README.md.temp >README.md
 
-node_modules/.bin/eslint -- README.md
+if [[ $module == esm ]] ; then
+  source=module
+else
+  source=script
+fi
+
+node_modules/.bin/eslint --parser-options "sourceType:$source" -- README.md
```
Comment (davidchambers) [inline bin/lint-readme]: For the benefit of future diffs, let's wrap the line above like so:

```bash
node_modules/.bin/eslint \
  --parser-options "sourceType:$source" \
  -- README.md
```

File: functions [hunk 5]
```diff
@@ -33,6 +33,7 @@ get() {
     closing-delimiter)      # shellcheck disable=SC2016
                             printf '```' ;;
     version-tag-prefix)     printf 'v' ;;
+    module-type)            printf 'commonjs' ;;
```
Comment (davidchambers) [inline functions:35]: Let's use the same order here as we do in the readme.

File: bin/lint-readme (hunk 4 above)
Comment (Avaq) [reply bin/lint-readme]: :zap:

File: functions (hunk 5 above)
Comment (Avaq) [reply functions:35]: :zap:

File: README.md [hunk 6]
```diff
@@ -95,6 +95,7 @@ Many variables have default values and are therefore optional.
 | `comment-prefix`              | `.`                   | The character which follows `//` to signify documentation to transcribe.      |
 | `opening-delimiter`           | `` ```javascript ``   | The opening delimiter of doctest blocks in the source files.                  |
 | `closing-delimiter`           | `` ``` ``             | The closing delimiter of doctest blocks in the source files.                  |
+| `module-type`                 | `commonjs`            | The type of module doctest should use. One of `commonjs`, `amd`, or `esm`.    |
```
Comment (davidchambers) [inline README.md]: Is it necessary to mention doctest here? This variable could also be used for other purposes.

File: bin/doctest [hunk 7]
```diff
@@ -11,12 +11,13 @@ set +f ; shopt -s nullglob
 files=($(get source-files))
 set -f ; shopt -u nullglob
 
+module="$(get module-type)"
 prefix="$(get comment-prefix)"
 opening="$(get opening-delimiter)"
 closing="$(get closing-delimiter)"
```
Comment (davidchambers) [inline bin/doctest:16]: Perhaps `get module-type` should come after `get closing-delimiter`. One day we could add linting to ensure that the variables mentioned in a script's documentation correspond to those it actually uses. This might be easier to do if the variables are listed in the same order in both places.

Comment (Avaq) [reply bin/doctest:16]: Currently I've put them in order of appearance in the script below.

File: README.md (hunk 6 above)
Comment (Avaq) [reply README.md]: But it's not. I think it's better to change the description once it's used for other purposes.

File: bin/doctest (hunk 7 above)
Comment (Avaq) [reply bin/doctest:16]: :zap:

File: README.md (hunk 6 above)
Comment (davidchambers) [reply README.md]: Okay. :)

File: functions [hunk 8]
```diff
@@ -71,3 +72,7 @@ pass() {
 fail() {
   echo $'\x1B[0;31m\xE2\x9C\x98\x1B[0m' "$@"
 }
+
+node_major_version() {
+  node -p process.versions.node | cut -d'.' -f1
```
Comment (davidchambers) [inline functions]: I don't know why it's so common to see option names and values squashed together, for `cut` in particular.

Let's give the arguments some breathing room:

```suggestion
  node --print process.versions.node | cut -d . -f 1
```

File: bin/doctest [hunk 9]
```diff
@@ -14,9 +14,15 @@ set -f ; shopt -u nullglob
...
+module="$(get module-type)"
+
+if [[ "$module" == esm ]] && [ "$(node_major_version)" -lt 9 ] ; then
```
Comment (davidchambers) [inline bin/doctest]:
```suggestion
if [[ "$module" == esm ]] && [[ "$(node_major_version)" -lt 9 ]] ; then
```

Comment (Avaq) [reply bin/doctest]: Oh! I thought `[[` doesn't support `-lt` for some reason.

File: bin/doctest [hunk 10]
```diff
@@ -14,9 +14,15 @@ set -f ; shopt -u nullglob
 prefix="$(get comment-prefix)"
 opening="$(get opening-delimiter)"
 closing="$(get closing-delimiter)"
+module="$(get module-type)"
+
+if [[ "$module" == esm ]] && [[ "$(node_major_version)" -lt 9 ]] ; then
+  echo 'Skipping ESM doctests on Node versions before 9.0.0'
```
Comment (davidchambers) [inline bin/doctest]:
```suggestion
  echo 'Skipping ESM doctests on Node.js version less than v9.0.0' >&2
```

Rationale:

  - [Node.js][1] seems to be the official project name.
  - The message should refer to a single run rather than to the policy, so _version_ rather than _versions_.
  - It's imprecise to refer to versions released _before_ a particular version, as it implies that all versions released after that version are supported. v8.15.1, for example, was released after v9.0.0 but does not support ESM.
  - The output should go to standard error for compatibility with scripts which process standard output. Although I don't recommend it, `npm run doctest | tail -n +6 | tr -d $'\n' | wc -c` could be used to report the number of doctests run.


[1]: https://en.wikipedia.org/wiki/Node.js

Comment (Avaq) [reply bin/doctest]: Great points all of them! :D

File: bin/doctest (hunk 9 above)
Comment (davidchambers) [reply bin/doctest]: My understand is that `[[` supports a superset of the `[` functionality. The main difference between the two is that `[[` is a language feature whereas `[` is a program:

```console
$ which -a [
/bin/[

$ which -a [[
```

By virtue of being a language feature `[[` gets to do things that are impossible for programs. As a result, `[[` has fewer traps than `[`.

File: package.json [hunk 11]
```diff
@@ -33,7 +33,7 @@
     "npm": ">=3.0.0"
   },
   "dependencies": {
-    "doctest": "0.16.x",
+    "doctest": "0.17.x",
```
Comment (davidchambers) [inline package.json]: Would it make sense for us to first make this change in a separate pull request?

Comment (Avaq) [reply package.json]: I think it'll be fine to rearrange the commits and leave it in this PR. Other than adding esm support, there is no pressing reason to update doctest.

Comment (Avaq) [reply package.json]: I force-pushed a new commit order

File: README.md (hunk 6 above)
Comment (davidchambers) [inline README.md]:
```suggestion
| `module-type`                 | `commonjs`            | The module system doctest should use (`amd`, `commonjs`, or `esm`).           |
```

**Comment by davidchambers** [comment]
I left one last suggestion. Aside from that, everything looks very good to me. Let's squash and merge. :)

**Review by davidchambers — APPROVED** [review]
